splunk_otel/profiling/__init__.py

```diff
@@ -0,0 +1,405 @@
+    )
+
+    profiling_logger = profiler.logger_provider.get_logger("otel.profiling", "0.1.0")
+
+    call_stack_interval_seconds = call_stack_interval_millis / 1e3
+    # In case the processing takes more than the given interval, default to the smallest allowed interval
+    min_call_stack_interval_seconds = 1 / 1e3
+    while profiler.running:
+        time_begin = time.time()
+        timestamp = int(time_begin * 1e9)
+
+        stacktraces = _collect_stacktraces(ignored_thread_ids)
+
+        log_record = _to_log_record(
+            stacktraces, timestamp, call_stack_interval_millis, options.resource
+        )
+
+        profiling_logger.emit(log_record)
+
+        processing_time = time.time() - time_begin
+        wait_for = max(
+            call_stack_interval_seconds - processing_time, min_call_stack_interval_seconds
+        )
+
+        with profiler.condition:
+            profiler.condition.wait(wait_for)
+
+    # Hack around batch log processor getting stuck on application exits when the profiling endpoint is not reachable
+    # Could be related: https://github.com/open-telemetry/opentelemetry-python/issues/2284
+    # pylint: disable-next=protected-access
+    profiler.exporter._shutdown = True
+    profiler.logger_provider.shutdown()
+    with profiler.condition:
+        profiler.condition.notify_all()
+
+
+def _wrapped_context_attach(wrapped, _instance, args, kwargs):
+    token = wrapped(*args, **kwargs)
+
+    maybe_context = args[0] if args else None
+
+    if maybe_context:
+        span = maybe_context.get(_SPAN_KEY)
+
+        if span:
+            thread_id = threading.get_ident()
+            _profiler.thread_states[thread_id] = (
+                span.context.trace_id,
+                span.context.span_id,
+            )
+
+    return token
+
+
+def _wrapped_context_detach(wrapped, _instance, args, kwargs):
+    token = args[0] if args else None
+
+    if token:
+        prev = token.old_value
+        thread_id = threading.get_ident()
+        if isinstance(prev, Context):
+            span = prev.get(_SPAN_KEY)
+
+            if span:
+                _profiler.thread_states[thread_id] = (
+                    span.context.trace_id,
+                    span.context.span_id,
+                )
+            else:
+                _profiler.thread_states[thread_id] = None
+        else:
+            _profiler.thread_states[thread_id] = None
+    return wrapped(*args, **kwargs)
+
+
+def _start_profiler_thread():
+    profiler_thread = threading.Thread(
+        name="splunk-otel-profiler", target=_profiler_loop, args=[_profiler], daemon=True
+    )
+    profiler_thread.start()
+
+
+def _force_flush():
+    if _profiler.logger_provider:
+        _profiler.logger_provider.force_flush()
+
+
+def _start_profiling(options):
+    if _profiler.running:
+        logger.warning("profiler already running")
+        return
+
+    _profiler.options = options
```

Comment:
If `Profiler` was an user-facing API, why not. But in this case it's only a private raw state holder (to refrain from using multiple globals and having to use `global X` when doing assignments). I think adding these functions would only cause unnecessary indirection when reading code, at the moment I don't see a reason to do it. If there's a good reason, sure - I might be missing something as well.

Reply:
Regarding the constructor: I want the default state to be uninitialized state, so I don't have to check for the existence of the profiler (e.g. `if _profiler and _profiler._logger_provider`) in each of the global-accessing functions (`stop_profiling`, `force_flush`, `start_profiling`). Could add a `start(options)` functions which initializes it completely, or do you mean making `options` optional in the constructor and initialize the remainder only if they are passed in?

Reply:
A couple of principles that I try to remember to use are:

* Minimize access to global variables. If you must access a global variable, do it once in a getter (and maybe once in a setter). Other than that, pass it around as a function argument (like [this code](https://github.com/signalfx/splunk-otel-python/pull/327/files#diff-4d67cd7b39484835d62dd8b53be7673692c73194cb05af6704eb6df534981445R257)). (Then almost all of your tests won't have to prepare/manipulate global state either.)
* With the exception of data classes, don't access a class's fields from outside the class. Instead encapsulate field access using methods.

A lot of the functions in this file are exposed as functions that modify global state, but look to me like they would be cohesive to Profiler as methods (because they read and write Profiler fields and perform related logic).

Regarding your last question above -- I don't really have a strong opinion, although I would prefer that the `Profiler` not be coupled to `_Options` at all and just be able to be initialized in a constructor or, if necessary, some other method(s) that sets its state after construction.

Reply:
Removed access to the `_profiler` global from `_start_profiler_thread` and `_encode_cpu_profile` and made `_wrapped_context_{attach,detach}` use a local profiler given to them via `_make_wrapped_context_{attach,detach}`.

The remainder of the functions having access to it are `_start_profiling`, `_force_flush`, `stop_profiling`.

They could have an optional profiler argument, defaulting to the global profiler, but why if this codepath is never used? If it is eventually needed for unit testing, it can be added, otherwise it's just dead code.

I'm not sure what a global `get_profiler` would get us over accessing `_profiler` directly. There is no locking involved and it would not make a difference when mocking. Especially if the function would look like:

```python
def get_profiler():
  return _profiler
```

Reply:
>A lot of the functions in this file are exposed as functions that modify global state, but look to me like they would be cohesive to Profiler as methods (because they read and write Profiler fields and perform related logic).

Which ones do you have in mind?

Reply:
Hey Siim, I have created a new PR against this PR to demonstrate what I'm proposing: https://github.com/signalfx/splunk-otel-python/pull/355

Reply:
Thanks, I don't mind, just some of the changes (such as `get_profiler`) make me question what do we gain from using a function when it doesn't do anything, but it definitely isn't worth the time to keep arguing. Can you mark it as ready? I can merge it then

Reply:
Feel free to remove the function -- I don't think it's that important.